Feature/Configuring PII /Anyonymization

Abhinegi2 opened the pull request:
closes:#2120

### Visible/Frontend Changes
- [x] Added a New tab to configure PII settings of Entity type.
- [x] Added checkbox to change the Anonymize settings of Entity.
- [x] Added Mat-Table to show the fields related to anonymized settings.
- [x] While switching between tabs PII changes will not lost until.

sleidig commented on src/app/core/admin/admin-entity-details/admin-entity-field/admin-entity-field.component.html:
```diff
@@ -145,7 +145,7 @@ <h2 mat-dialog-title i18n>Configure Field "{{ entitySchemaField.label }}"</h2>
         label="Advanced Options & Validation [COMING SOON]"
         i18n-label
-        [disabled]="true"
+        [disabled]="false"
```
```suggestion
```

sleidig commented:
- [ ] "has PII" checkbox needs to be saved and loaded from entity

sleidig commented on src/app/core/admin/admin-entity/admin-entity-general-settings/admin-entity-general-settings.component.html:
```diff
@@ -70,8 +74,53 @@ <h2 i18n>General Settings of "{{ entityConstructor.label }}" Records</h2>
+      [disabled]="false"
+    >
+      <div class="grid-layout margin-top-regular">
+        <mat-checkbox [checked]="showTable" (change)="toggleTable($event)"
```
add `i18n` here and for other parts of template where missing

sleidig commented on src/app/core/admin/admin-entity/admin-entity.component.ts:
```diff
@@ -86,6 +85,10 @@ export class AdminEntityComponent implements OnInit {
   }
 
+  updateCheckboxState(checkboxStatus: boolean) {
+    this.checkboxState = checkboxStatus;
```
why do you need to interact with checkbox PII status in the parent component? This should be completely encapsulated within AdminGeneralSettingsComponent only and saved via the configEntitySettings together with other config.

--> see/map to EntityConfig.hasPII

Abhinegi2 replied:
Previously  am using a custom checkbox to open the PII configuration, But now I know that we can use has PII from entity, So will change all occurrences.

sleidig commented:
@Abhinegi2 , experimenting with the UI I have decided on some adjustment and still see some polishing to do:
- [x] removed pagination, to give user a full, seamless overview of all fields
- [ ] limit the height, so that the scroll is within the "Configure PII" tab, not for the whole screen if many fields are listed
- [ ] fields need to display the field.label in the table, not the key/id

Abhinegi2 commented:
> @Abhinegi2 , experimenting with the UI I have decided on some adjustment and still see some polishing to do:
> 
> * [x]  removed pagination, to give user a full, seamless overview of all fields
> * [ ]  limit the height, so that the scroll is within the "Configure PII" tab, not for the whole screen if many fields are listed
> * [ ]  fields need to display the field.label in the table, not the key/id

sure, will do these adjustment

Abhinegi2 commented:
> @Abhinegi2 , experimenting with the UI I have decided on some adjustment and still see some polishing to do:
> 
> * [x]  removed pagination, to give user a full, seamless overview of all fields
> * [x]  limit the height, so that the scroll is within the "Configure PII" tab, not for the whole screen if many fields are listed
> * [x]  fields need to display the field.label in the table, not the key/id

Hi @sleidig I have implemented the following feedback can you please confirm whether UI is looking good or not - 
1. Added height limit on PII fields.
2. Display the field name instead of the field key.

sleidig commented:
@Abhinegi2 , the scroll is still not working very well:
![image](https://github.com/Aam-Digital/ndb-core/assets/1682541/46d2b25f-3c54-4de3-bd59-00fc0b19cd4a)


I would like to 
- always have only on scroll bar (only scroll within the tabs' contents, i.e. the inner scrollbar)
- use the full height of the (remaining) screen before having to scroll (i.e. avoid the white space at bottom currently)

sleidig commented on src/app/core/admin/admin-entity/admin-entity-general-settings/admin-entity-general-settings.component.ts:
```diff
@@ -38,19 +48,36 @@ import { StringDatatype } from "../../../basic-datatypes/string/string.datatype"
+  @Input() showPIIDetails: boolean;
+  entitySchemaField: EntitySchemaField;
+  dataSource: MatTableDataSource<any>;
+  anonymizOptionList: string[] = ["Retain", "Partially Anonymize", "Remove"];
```
Let's add the tooltips here: `anonymizOptionList: { label: string; tooltip: string }[] =`
(calling methods in the template is not ideal, as they are executed on every change detection cycle of angular. And it's easier to maintain the texts, if they are all in one place here).

sleidig replied:
i.e. remove the getTooltip method for this then

sleidig replied:
actually, we can also hard code the three options including tooltip directly in the template html. Anyway these are not dynamically generated currently.

sleidig commented on src/app/core/admin/admin-entity/admin-entity-general-settings/admin-entity-general-settings.component.ts:
```diff
@@ -38,19 +48,36 @@ import { StringDatatype } from "../../../basic-datatypes/string/string.datatype"
+  allPIIFields: any[];
+  @Input() showPIIDetails: boolean;
+  entitySchemaField: EntitySchemaField;
+  dataSource: MatTableDataSource<any>;
```
more specific naming, please: e.g. fieldAnonymizationDataSource

sleidig commented on src/app/core/admin/admin-entity/admin-entity-general-settings/admin-entity-general-settings.component.ts:
```diff
@@ -38,19 +48,36 @@ import { StringDatatype } from "../../../basic-datatypes/string/string.datatype"
   @Input() generalSettings: EntityConfig;
+  allPIIFields: any[];
+  @Input() showPIIDetails: boolean;
+  entitySchemaField: EntitySchemaField;
```
this seems to never get assigned to anything?

sleidig commented on src/app/core/admin/admin-entity/admin-entity-general-settings/admin-entity-general-settings.component.ts:
```diff
@@ -65,18 +92,92 @@ export class AdminEntityGeneralSettingsComponent implements OnInit {
+    }
+  }
+
+  selectionOnChange(value: any, selectFielddata: any) {
```
clearer naming please, e.g. changeFieldAnonymization

also, can you figure out the actual types? `any` is something we try to avoid at all cost.

sleidig commented on src/app/core/admin/admin-entity/admin-entity-general-settings/admin-entity-general-settings.component.ts:
```diff
@@ -65,18 +92,92 @@ export class AdminEntityGeneralSettingsComponent implements OnInit {
+    } else if ((selectFielddata.anonymize = "Remove")) {
+      selectedFielddetails.anonymize = undefined;
+    }
+    const updatedEntitySchemaMerged = Object.assign(
+      { _isCustomizedField: true },
+      this.entitySchemaField,
+      selectedFielddetails,
+    );
```
It should be enough to simply do `this.entityConstructor.schema.get(selectedFieldId).anonymize = newAnonymizeValue` 

also, you can use `<mat-option [value]="..."` to set the exact key (e.g. "retain" instead of "Retain") and avoid having to map it here in the method again.

sleidig commented on src/app/core/admin/admin-entity/admin-entity.component.scss:
```diff
@@ -26,4 +26,9 @@
     color: colors.$muted;
   }
+
+  &.general-mode {
+    overflow-y: hidden;
+  }
```
This is not a feasible solution as it will make parts of the UI completely unreachable on smaller windows.

sleidig replied:
I have removed this. If we cannot figure out the correct height for the mat-tab-group , it's better to scroll the full part rather than hide completely

Abhinegi2 commented:
> @Abhinegi2 , the validator PR has been merged. Can you please replace the simple input field in the AdminEntityField with the new dropdown component here also

sure working on this part
